### PR TITLE
Don't ask for an invalid group in an HRR

### PR DESCRIPTION
**Pull request by mattcaswell**
If the client sends us a group in a key_share that is in our
supported_groups list but is otherwise not suitable (e.g. not compatible
with TLSv1.3) we reject it. We should not ask for that same group again
in a subsequent HRR.

Fixes https://github.com/openssl/openssl/issues/21157

### CODE REVIEW

**Review by t8m — APPROVED** [review]

File: test/recipes/70-test_tls13hrr.t [hunk 1]
```diff
@@ -131,16 +150,25 @@ sub hrr_filter
...
-    # The server will always pick TLS_AES_256_GCM_SHA384
-    my @ciphersuites = (TLSProxy::Message::CIPHER_TLS13_AES_128_GCM_SHA256);
-    $ch1->ciphersuite_len(2 * scalar @ciphersuites);
-    $ch1->ciphersuites(\@ciphersuites);
+    if ($testtype == CHANGE_CH1_CIPHERSUITE) {
+        # The server will always pick TLS_AES_256_GCM_SHA384
+        my @ciphersuites = (TLSProxy::Message::CIPHER_TLS13_AES_128_GCM_SHA256);
+        $ch1->ciphersuite_len(2 * scalar @ciphersuites);
+        $ch1->ciphersuites(\@ciphersuites);
+    } else {
```
Comment (tmshort) [inline test/recipes/70-test_tls13hrr.t:164]: Nit: This should probably be `if ($testtype == INVALID_GROUP) {`, with or without an `else`.

This future-proofs the test a bit.

File: ssl/statem/extensions.c [hunk 2]
```diff
@@ -1449,7 +1449,11 @@ static int final_key_share(SSL_CONNECTION *s, unsigned int context, int sent)
                     group_id = pgroups[i];
 
                     if (check_in_list(s, group_id, clntgroups, clnt_num_groups,
-                                      1))
+                                      1)
```
Comment (tmshort) [inline ssl/statem/extensions.c:1452]: Does line-length allow for this to be put with the previous line?

**Review by tmshort — APPROVED** [review]
Just a couple minor things that don't block approval.

File: ssl/statem/extensions.c (hunk 2 above)
Comment (mattcaswell) [reply ssl/statem/extensions.c:1452]: No - line-length prevents this

**Comment by mattcaswell** [comment]
Fix up pushed to address the feedback. @t8m and @tmshort - please can you reconfirm?

**Review by t8m — APPROVED** [review]

**Review by tmshort — APPROVED** [review]

**Comment by tmshort** [comment]
Still good, though the `1)` on it's own line is ugly!

**Comment by mattcaswell** [comment]
Pushed to master/3.1/3.0. Thanks.

**Pull request closed without merging**